xarray/core/resample_cftime.py

```diff
@@ -0,0 +1,250 @@
+                                     Day, MonthEnd, YearEnd,
+                                     CFTIME_TICKS, to_offset)
+import datetime
+import numpy as np
+import pandas as pd
+
+
+class CFTimeGrouper(object):
+    """This is a simple container for the grouping parameters that implements a
+    single method, the only one required for resampling in xarray.  It cannot
+    be used in a call to groupby like a pandas.Grouper object can."""
+
+    def __init__(self, freq, closed, label, base):
+        self.freq = to_offset(freq)
+        self.closed = closed
+        self.label = label
+        self.base = base
+
+    def first_items(self, index):
+        """Meant to reproduce the results of the following
+
+        grouper = pandas.Grouper(...)
+        first_items = pd.Series(np.arange(len(index)), index).groupby(grouper).first()
+
+        with index being a CFTimeIndex instead of a DatetimeIndex.
+        """
+        datetime_bins, labels = _get_time_bins(index, self.freq, self.closed,
+                                               self.label, self.base)
+
+        # check binner fits data
+        if index[0] < datetime_bins[0]:
+            raise ValueError("Value falls before first bin")
+        if index[len(index) - 1] > datetime_bins[len(datetime_bins) - 1]:
+            raise ValueError("Value falls after last bin")
+
+        integer_bins = np.searchsorted(index, datetime_bins, side=self.closed)[
+                       :-1]
+        if len(integer_bins) < len(labels):
+            labels = labels[:len(integer_bins)]
+        first_items = pd.Series(integer_bins, labels)
+
+        # Mask duplicate values with NaNs, preserving the last values
+        non_duplicate = ~first_items.duplicated('last')
+        return first_items.where(non_duplicate)
+
+
+def _get_time_bins(index, freq, closed, label, base):
+    """This is basically the same with the exception of the call to
+    _adjust_bin_edges."""
+
+    if not isinstance(index, CFTimeIndex):
+        raise TypeError('index must be a CFTimeIndex, but got '
+                        'an instance of %r' % type(index).__name__)
+    if len(index) == 0:
+        datetime_bins = labels = CFTimeIndex(data=[], name=index.name)
+        return datetime_bins, labels
+
+    first, last = _get_range_edges(index.min(), index.max(), freq,
+                                   closed=closed,
+                                   base=base)
+    datetime_bins = labels = cftime_range(freq=freq,
+                                          start=first,
+                                          end=last,
+                                          name=index.name)
+
+    datetime_bins = _adjust_bin_edges(datetime_bins, freq, closed, index)
+
+    if closed == 'right':
+        if label == 'right':
+            labels = labels[1:]
+    elif label == 'right':
+        labels = labels[1:]
+
+    if index.hasnans:  # cannot be true since CFTimeIndex does not allow NaNs
```

Comment:
Thanks for noting this -- maybe instead of adding this code just yet, insert a `TODO` comment:
```python
# TODO: when CFTimeIndex supports missing values, if the reference index
# contains missing values, insert the appropriate NaN value at the
# beginning of the datetime_bins and labels indexes.
```